skxray/correlation.py

```diff
@@ -0,0 +1,299 @@
+    ----------
+    num_levels : int
+        number of levels of multiple-taus
+
+    num_bufs : int, even
+        number of channels or number of buffers in
+        auto-correlators normalizations (must be even)
+
+    num_qs : int
+        number of region of interests(roi's)
+
+     ring_inds : ndarray
+        indices of the required rings
+
+    pixel_list : ndarray
+        pixel indices for the required roi's
+
+    q_inds : ndarray
+        indices of the required roi's
+
+    img_stack : ndarray
+        intensity array of the images
+        dimensions are: [num_img][num_rows][num_cols]
+
+    Returns
+    -------
+    g2 : ndarray
+        matrix of one-time correlation
+
+    lag_steps : ndarray
+        delay or lag steps for the multiple tau analysis
+
+    elapsed_times : float
+        elapsed time for auto correlation (in seconds)
+
+    Notes
+    -----
+    In order to calculate correlations for delays, images must be
+    keep for upto the maximum delay. These are stored in the array
+    buf. This algorithm only keeps number of buffers and delays but
+    several levels of delays number of levels are kept in buf. Each
+    level has twice the delay times of the next lower one. To save
+    needless copying, of cyclic storage of images in buf is used.
+
+    References: text [1]_
+
+    .. [1] D. Lumma, L. B. Lurio, S. G. J. Mochrie and M. Sutton,
+        "Area detector based photon correlation in the regime of
+        short data batches: Data reduction for dynamic x-ray
+        scattering," Rev. Sci. Instrum., vol 70, p 3274-3289, 2000.
+
+    """
+    # number of pixels in required roi's, dimensions are : [num_qs]X1
+    num_pixels = np.bincount(q_inds, minlength=(num_qs+1))
+    num_pixels = num_pixels[1:]
+
+    for item in num_pixels:
+        if item == 0:
+            raise ValueError("Number of pixels of the required roi's"
+                             " cannot be zero,"
+                             " num_pixels = {0}".format(num_pixels))
+
+    # matrix of auto-correlation function without normalizations
+    G = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                 dtype=np.float64)
+    # matrix of past intensity normalizations
+    IAP = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                   dtype=np.float64)
+    # matrix of future intensity normalizations
+    IAF = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                   dtype=np.float64)
+
+    # matrix of one-time correlation for required roi's
+    g2 = np.zeros((num_levels, num_qs), dtype=np.float64)
+
+    # correlation for delays, images must be keep for up to maximum
+    # delay in buf
+    buf = np.zeros((num_levels, num_bufs, np.sum(num_pixels)),
+                   dtype=np.float64)
+
+    cts = np.zeros(num_levels)
+    cur = np.ones((num_levels), dtype=np.int64)
+
+    num = np.array(np.zeros(num_levels), dtype=np.int64)
+
+    # number of frames(images)
+    num_frames = img_stack.shape[0]
```

Comment:
do this as `len(img_stack)` so that we don't have to have all of the images in core to run this code.